api/APIValidation.ts

```diff
@@ -0,0 +1,84 @@
+import { ZodError, ZodTypeAny, z } from "zod";
+import { addLogHandler, consoleLogHandler, logger } from "../logging";
+import { APIMiddleware, TiFAPIInputContext, TiFAPIResponse } from "./TransportTypes";
+
+const log = logger("tif.api.validation")
+addLogHandler(consoleLogHandler())
+
+const zodValidation = async <T>(data: T, schema: ZodTypeAny): Promise<T | "failure"> => {
+  try {
+      return await schema.parseAsync(data);
+  } catch (e) {
+      if (e instanceof ZodError) {
+          log.error("Zod Schema Error Message", {
+              zodError: JSON.parse(e.message)
+          });
+      } else {
+          log.error(e);
+      }
+      return "failure";
+  }
+};
+
+export type ValidationResult = "invalid-request" | "unexpected-response" | "invalid-response" | "passed"
+
+export type ValidationResultParser = 
+  ((_: {validationStatus: Extract<ValidationResult, "invalid-request">, request: TiFAPIInputContext<any>} 
+    | {validationStatus: Extract<ValidationResult, "unexpected-response" | "invalid-response">, response: TiFAPIResponse<any>} 
+    | {validationStatus: Extract<ValidationResult, "passed">, response: TiFAPIResponse<any>}
+  ) => (TiFAPIResponse<any>))
```

Comment:
Do we need `Extract` here? It seems a bit redundant. Also I think it's fine if all cases get access to the input context, but it seems that the response is only omitted from the `"invalid-request"` case so I think we can simplify this type quite a bit.